src/rt/ds/stack.h

```diff
@@ -7,128 +7,306 @@
+    void dealloc(Alloc* alloc)
     {
-      auto local_block = block;
-      while (local_block)
+      auto local_block = get_block(index);
+      while (local_block != &null_block)
       {
-        auto prev = local_block->prev;
+        auto prev = get_block(local_block->prev);
         alloc->template dealloc<sizeof(Block)>(local_block);
         local_block = prev;
       }
-      if (backup != nullptr)
-        alloc->template dealloc<sizeof(Block)>(backup);
     }
 
+    /// returns true if this stack is empty
     ALWAYSINLINE bool empty()
     {
-      return block == nullptr;
+      return index == null_index;
     }
 
-    ALWAYSINLINE T peek()
+    /// Return the top element of the stack without removing it.
+    ALWAYSINLINE T* peek()
     {
       assert(!empty());
-      return block->data[index - 1];
+      return *index;
     }
 
-    ALWAYSINLINE T pop()
+    /// Call this to pop an element from the stack.
+    ALWAYSINLINE T* pop(Alloc* alloc)
     {
       assert(!empty());
+      if (!is_empty(index - 1))
+      {
+        auto item = peek();
+        index--;
+        return item;
+      }
 
-      index--;
-      T item = block->data[index];
-
-      if (index == 0)
-        pop_slow_path();
-
-      return item;
+      return pop_slow(alloc);
     }
 
-    ALWAYSINLINE void push(T item)
+    /// Call this to push an element onto the stack.
+    ALWAYSINLINE void push(T* item, Alloc* alloc)
     {
-      if (index < STACK_COUNT)
+      if (!is_full(index))
       {
-        block->data[index] = item;
         index++;
+        *index = item;
+        return;
       }
-      else
-      {
-        push_slow_path(item);
-      }
+
+      push_slow(item, alloc);
     }
 
-    template<void apply(T t)>
+    /// For all elements of the stack
+    template<void apply(T* t)>
     void forall()
     {
-      Block* curr = block;
-      size_t i = index;
+      T* curr = index;
 
-      while (curr != nullptr)
+      while (curr != null_index)
       {
         do
         {
-          i--;
-          apply(curr->data[i]);
-        } while (i > 0);
+          apply(*curr);
+          curr--;
+        } while (is_empty(curr));
 
-        curr = curr->prev;
-        i = STACK_COUNT;
+        curr = get_block(curr)->prev;
       }
     }
 
   private:
-    void pop_slow_path()
+    /// Slow path for push, performs a push, when allocation is required.
+    void push_slow(T* item, Alloc* alloc)
     {
-      Block* prev = block->prev;
+      assert(is_full(index));
 
-      if (backup != nullptr)
-        alloc->template dealloc<sizeof(Block)>(backup);
+      Block* block = (Block*)alloc->template alloc<sizeof(Block)>();
+      T** iter = (T**)block;
+      assert(is_empty(iter));
+      auto next = get_block(iter);
 
-      backup = block;
-      block = prev;
-      index = STACK_COUNT;
+      assert(index != (T**)&null_block);
+      next->prev = index;
+      index = &(next->data[0]);
+      next->data[0] = item;
```

Comment:
I've just realized block and next are the same, so actually this could be simplified quite a bit.
```c++
Block* next = (Block*)alloc->template alloc<sizeof(Block)>();
next->prev = index;
next->data[0] = item;
index = &(next->data[0]);
```

Reply:
That was done to check alignment assumptions, but I can do that much more cleanly